### PR TITLE
Add initial support for ICUs in template pipeline

### PR DESCRIPTION
**Pull request by mmalerba**
See individual commit messages for details.

### CODE REVIEW

File: packages/compiler/src/template/pipeline/src/ingest.ts [hunk 1]
```diff
@@ -383,6 +385,16 @@ function ingestDeferBlock(unit: ViewCompilationUnit, deferBlock: t.DeferredBlock
   unit.create.push(deferOnOp);
 }
 
+function ingestIcu(unit: ViewCompilationUnit, icu: t.Icu) {
+  if (icu.i18n instanceof i18n.Message) {
+    const xref = unit.job.allocateXrefId();
+    unit.create.push(ir.createIcuOp(xref, icu.i18n, null!));
```
Comment (dylhunn) [inline packages/compiler/src/template/pipeline/src/ingest.ts:391]: Just a heads up, i18n spans have tests in the source spans unit test folder. I don't blame you for not wanting to deal with that at this point though.

File: packages/compiler/src/template/pipeline/ir/src/enums.ts [hunk 2]
```diff
@@ -426,3 +426,18 @@ export enum BindingKind {
    */
   Animation,
 }
+
+/**
+ * Enumeration of possible times i18n params can be resolved.
+ */
+export enum I18nParamResolutionTime {
+  /**
+   * Param is resolved at message creation time.
```
Comment (dylhunn) [inline packages/compiler/src/template/pipeline/ir/src/enums.ts:435]: Can you add some details about why/which placeholders get resolved at creation vs update time? Or is it just truly arbitrary?

File: packages/compiler/src/template/pipeline/src/phases/wrap_icus.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ir from '../../ir';
+import {CompilationJob} from '../compilation';
+
+/**
+ * Wraps ICUs that do not already belong to an i18n block in a new i18n block.
+ */
+export function phaseWrapIcus(job: CompilationJob): void {
```
Comment (dylhunn) [inline packages/compiler/src/template/pipeline/src/phases/wrap_icus.ts:15]: Makes sense. Weird `i18n` is not required though.

**Review by dylhunn — APPROVED** [review]

File: packages/compiler/src/template/pipeline/ir/src/enums.ts (hunk 2 above)
Comment (mmalerba) [reply packages/compiler/src/template/pipeline/ir/src/enums.ts]: 👍

File: packages/compiler/src/template/pipeline/src/ingest.ts (hunk 1 above)
Comment (mmalerba) [reply packages/compiler/src/template/pipeline/src/ingest.ts:391]: Yeah, I was thinking maybe I could do i18n (except source maps) and you could do source maps (except i18n) and then we can team up and use our combined knowledge to take down the combo

File: packages/compiler/src/template/pipeline/src/phases/wrap_icus.ts (hunk 3 above)
Comment (mmalerba) [reply packages/compiler/src/template/pipeline/src/phases/wrap_icus.ts:15]: I guess its slightly more ergonomic for users to just be able to throw an ICU in without needing to add an attribute 🤷

File: packages/compiler/src/template/pipeline/src/ingest.ts (hunk 1 above)
Comment (dylhunn) [reply packages/compiler/src/template/pipeline/src/ingest.ts:391]: Sounds scary! Non-i18n source maps are more or less done, so we'll team up when you're ready and find out how they interact... 👻

**Pull request closed without merging**